### DISCUSSION TITLE
Raspberry Pi Vulkan rendering speed

### DISCUSSION THREAD

**Discussion by DaveAllerton** [vsg-dev/VulkanSceneGraph#1022] [Q&A]
Does the warning below imply that my RPi 4B is not correctly configured for Vulkan? I believe this is the RPi Vulkan driver produced by Igalia. 18 Hz seems a very slow frame rate for a simple example.
I get the same warning with vsgdraw and vsgviewer, even though they run. With vsgdraw, should I see two rotating terrains? 

~/Projects/vsgTutorial>vsgdynamictexture
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
Average frame rate = 18.4578

**Comment by robertosfield** [comment]
18fps seems low for a simple test.

vsgdraw is an implementation of the Vulkan Tutorial so is just two textured quads, it just happens to have the Ed Levin Park terrain texture.

**Reply by DaveAllerton** [reply]
Vkcube rotates at a reasonable rate but I don't know the frame rate.

**Comment by timoore** [comment]
I'm not Raspberry Pi user, but it sounds to me as if you are still not using the hardware GPU, or the driver is advertising both the hardware renderer and a software alternative and VSG is choosing the wrong one.

What's the output of `vsgdeviceselection --list`?

**Comment by DaveAllerton** [comment]
~/dja/cgitest>vsgdeviceselection --list
vkEnumerateInstanceVersion() 4202658
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.

physicalDevices.size() = 2
    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0xf78213a0) V3D 4.2, deviceType = 1, apiVersion = 4194459, driverVersion = 83898373
    QueueFamilyProperties 1
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER, queueCount = 1, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}

    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0xf78217e0) llvmpipe (LLVM 11.0.1, 128 bits), deviceType = 4, apiVersion = 4194306, driverVersion = 1
    QueueFamilyProperties 1
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER, queueCount = 1, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}

I think the current RPi Vulkan driver was produced by Igalia and is 1.2 conformant.
I'm not sure itr's fair to compare the RPi Broadcom GPU with an nVidia card, although if I have done something stupid with my installation which is slowing the graphics down, it would be good to fix the problem. I have now got my installation running the vsgVulkan examples (vsgdraw, vsgviewer and vsgdynamic texture), vsgxconv and the HelloWorld.

**Reply by robertosfield** [reply]
The VSG by default will select the vk/vsg::PhysicalDevice that is hardware accelerated but you can manually select vsg::PhysicalDevice that logical vsg::Device is created with. The vsg::Device is essentially the handle to the driver that your application/VSG will use when doing all work. In your case I expect the V3D 4.2 driver is being used.  The llvmpipe is the software driver, but even it performs quite well on my AMD 5700G CPU on my desktop system.  

On my system this is what I see:

~~~ sh
$ vsgdeviceselection --list
vkEnumerateInstanceVersion() 4206860
VK_API_VERSION = 1.3.268.0

physicalDevices.size() = 2
    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7fe26f9e5c38) AMD Unknown (RADV RENOIR), deviceType = 1, apiVersion = 1.3.238, driverVersion = 23.0.4
    QueueFamilyProperties 2
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER | PARSE_BINDING, queueCount = 1, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
        VkQueueFamilyProperties[1] queueFlags = COMPUTE | TRANSFER | PARSE_BINDING, queueCount = 4, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}

    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7fe26f9e60a8) llvmpipe (LLVM 15.0.7, 256 bits), deviceType = 4, apiVersion = 1.3.238, driverVersion = 0.0.1
    QueueFamilyProperties 1
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER, queueCount = 1, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
~~~

And if I select the first PhysicalDevice using select:

~~~ sh
$ vsgdeviceselection --select 0
vkEnumerateInstanceVersion() 4206860
VK_API_VERSION = 1.3.268.0
Selected vsg::PhysicalDevice ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7f5d20c6ac38) AMD Unknown (RADV RENOIR) deviceType = 1
~~~

Selecting the software driver:
~~~ sh
$ vsgdeviceselection --select 1
vkEnumerateInstanceVersion() 4206860
VK_API_VERSION = 1.3.268.0
Selected vsg::PhysicalDevice ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7f1ea3a080a8) llvmpipe (LLVM 15.0.7, 256 bits) deviceType = 4
~~~
